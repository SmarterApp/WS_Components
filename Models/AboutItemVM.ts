import * as Rubric from '../PageTabs/Rubric'
import * as ItemCardViewModel from './ItemCardViewModel'
import * as API from '../Models/ApiModels'

export const ScoreSearchClient = (params: { bankKey: number, itemKey: number }) => API.get<AboutThisItem>("/api/aboutItem", params)

export interface AboutThisItem {
    rubrics: Rubric.Rubric[];
    itemCardViewModel: ItemCardViewModel.ItemCardViewModel;
    depthOfKnowledge: string;
    targetDescription: string;
    commonCoreStandardsDescription: string;
    educationalDifficulty: string;
    evidenceStatement: string;
}