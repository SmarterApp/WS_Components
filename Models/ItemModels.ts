import * as $ from 'jquery';
import { GradeLevels } from './GradeLevels';
import * as API from './ApiModels';
import * as ItemCardViewModel from './ItemCardViewModel';

export interface Subject {
    code: string;
    label: string;
}

export interface TechType extends Subject { }

export interface FilterOptions {
    subjects: Subject[];
    grades: GradeLevels[];
    techTypes: TechType[];
}

export interface ItemFilter extends FilterOptions { };