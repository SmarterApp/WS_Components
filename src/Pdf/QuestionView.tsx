import * as React from "react";
import { ItemView } from "./ItemView";
import { EvidenceStatement } from "./EvidenceStatement";
import { ItemCardTable } from "../ItemCard/ItemCardTable";
import { RubricTable } from "../Rubric/RubricTable";
import { QuestionModel } from "./PdfModels";
import { AboutItemModel } from "../AboutItem/AboutItemModels";

export interface QuestionViewProps {
  question: QuestionModel;
  displayScoreInfo: boolean;
}

export class QuestionView extends React.Component<QuestionViewProps, {}> {
  renderRubric(data?: AboutItemModel) {
    if (!this.props.displayScoreInfo) {
      return undefined;
    }
    let rubric: JSX.Element | undefined;
    if (data && data.sampleItemScoring && data.sampleItemScoring.rubrics) {
      const entries = data.sampleItemScoring.rubrics
        .map(r => r.rubricEntries)
        .reduce((acc, re) => acc.concat(re), []);
      const samples = data.sampleItemScoring.rubrics
        .map(r => r.samples)
        .reduce((acc, sample) => acc.concat(sample), []);

      if (entries.length > 1 || samples.length > 1) {
        rubric = <RubricTable rubrics={data.sampleItemScoring.rubrics} />;
      } else {
        const entriesJsx =
          samples.length > 0 ? (
            <div>
              <b>Rubric:</b> ({entries[0].scorepoint}point) {entries[0].value}
            </div>
          ) : (
            undefined
          );
        const samplesJsx =
          samples.length > 0 ? (
            <div>
              <b>Exemplar:</b> {samples[0]}
            </div>
          ) : (
            undefined
          );

        rubric = (
          <div>
            {samplesJsx} {entriesJsx}
          </div>
        );
      }
    }

    return rubric ? rubric : undefined;
  }

  render() {
    const data = this.props.question.data;

    const evidenceStatement =
      data && data.evidenceStatement ? (
        <EvidenceStatement statement={data.evidenceStatement} />
      ) : (
        undefined
      );

    const dataTable =
      data && data.itemCardViewModel ? (
        <ItemCardTable card={data.itemCardViewModel} />
      ) : (
        undefined
      );

    const key =
      data &&
      data.sampleItemScoring &&
      data.sampleItemScoring.answerKey &&
      this.props.displayScoreInfo ? (
        <div>
          <b>Key:</b> {data.sampleItemScoring.answerKey}
        </div>
      ) : (
        undefined
      );

    return (
      <div className="item">
        <p className="question-title">
          Question #{this.props.question.questionNumber}
        </p>
        {dataTable}
        {evidenceStatement}
        <ItemView view={this.props.question.view} />
        {key}
        {this.renderRubric(data)}
      </div>
    );
  }
}
