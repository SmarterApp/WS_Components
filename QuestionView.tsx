import * as React from 'react';
import { Question } from '../Models';
import { ItemViewComponent } from './ItemView';
import { EvidenceStatement } from './EvidenceStatement';
import { QuestionDataTable } from './QuestionDataTable';

interface Props {
    question: Question
}

export class QuestionView extends React.Component<Props, {}> {
    render() {
        const data = this.props.question.data;

        const evidenceStatement = data && data.evidenceStatement
            ? <EvidenceStatement statement={data.evidenceStatement} />
            : null;
        
        const dataTable = data && data.itemCardViewModel
            ? <QuestionDataTable tableData={data.itemCardViewModel} />
            : null;
        
        return (
            <div className='item'>
                <h2>Question #{this.props.question.questionNumber}</h2>
                {dataTable}
                {evidenceStatement}
                <ItemViewComponent view={this.props.question.view} />
            </div>
        );
    }
}