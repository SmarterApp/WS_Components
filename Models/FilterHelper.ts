import { ItemCardViewModel } from "../Models/ItemCardViewModel";
import { ItemFilter, FilterOptions } from "../Models/ItemModels";
import * as GradeLevels from '../Models/GradeLevels';
import { parseQueryString } from "../Models/ApiModels";
import { AdvancedFilterCategory, AdvancedFilters, OptionType, AdvancedFilterOption } from "../filter/AdvancedFilterModel";

export class FilterHelper {

    //TODO: Get this from the server
    static getFilterOptions(): AdvancedFilters  {
        const subjectsFilterOptions: AdvancedFilterOption[] = [{
            label: "Mathematics",
            key: "MATH",
            isSelected: false,
            type: OptionType.button
        },
        {
            label: "English",
            key: "ELA",
            isSelected: false,
            type: OptionType.button
        }];

        const gradesFilterOptions: AdvancedFilterOption[] = [{
            label: GradeLevels.caseToString(GradeLevels.GradeLevels.Elementary),
            key: String(GradeLevels.GradeLevels.Elementary),
            isSelected: false,
            type: OptionType.button
        }, {
            label: GradeLevels.caseToString(GradeLevels.GradeLevels.Grade3),
            key: String(GradeLevels.GradeLevels.Grade3),
            isSelected: false,
            type: OptionType.button
        }, {
            label: GradeLevels.caseToString(GradeLevels.GradeLevels.Grade4),
            key: String(GradeLevels.GradeLevels.Grade4),
            isSelected: false,
            type: OptionType.button
        }, {
            label: GradeLevels.caseToString(GradeLevels.GradeLevels.Grade5),
            key: String(GradeLevels.GradeLevels.Grade5),
            isSelected: false,
            type: OptionType.button
        }, {
            label: GradeLevels.caseToString(GradeLevels.GradeLevels.Grade6),
            key: String(GradeLevels.GradeLevels.Grade6),
            isSelected: false,
            type: OptionType.button
        }, {
            label: GradeLevels.caseToString(GradeLevels.GradeLevels.Grade7),
            key: String(GradeLevels.GradeLevels.Grade7),
            isSelected: false,
            type: OptionType.button
        }, {
            label: GradeLevels.caseToString(GradeLevels.GradeLevels.Grade8),
            key: String(GradeLevels.GradeLevels.Grade8),
            isSelected: false,
            type: OptionType.button
        }, {
            label: GradeLevels.caseToString(GradeLevels.GradeLevels.High),
            key: String(GradeLevels.GradeLevels.High),
            isSelected: false,
            type: OptionType.button
        }];

        const techTypesFilterOptions: AdvancedFilterOption[] = [{
            label: "CAT",
            key: "CAT",
            isSelected: false,
            type: OptionType.button
        }, {
            label: "Performance Items",
            key: "PT",
            isSelected: false,
            type: OptionType.button
        }];

        const subjects: AdvancedFilterCategory = {
            disabled: false,
            isMultiSelect: false,
            label: "Subjects",
            helpText: "Subjects HelpText here.",
            filterOptions: [...subjectsFilterOptions],
            displayAllButton: true
        }

        const grades: AdvancedFilterCategory = {
            disabled: false,
            isMultiSelect: false,
            label: "Grade",
            helpText: "Grade HelpText here.",
            filterOptions: [...gradesFilterOptions],
            displayAllButton: true
        }

        const techTypes: AdvancedFilterCategory = {
            disabled: false,
            isMultiSelect: false,
            label: "TechType",
            helpText: "TechType HelpText here.",
            filterOptions: [...techTypesFilterOptions],
            displayAllButton: false
        }

        const value: AdvancedFilters = {subjects, grades, techTypes};
        
        return value;
    }

    static filter(itemCards: ItemCardViewModel[], filter: AdvancedFilterCategory[]): ItemCardViewModel[] {
        const grades = filter.find(afc => afc.label === "Grade");
        const subjects = filter.find(afc => afc.label === "Subjects");
        const techTypes = filter.find(afc => afc.label === "TechType");

        if (grades && grades.filterOptions) {
            let selectedGrades = GradeLevels.GradeLevels.NA;

            grades.filterOptions.forEach(gradeFilter => {
                if (gradeFilter.isSelected) {
                    selectedGrades = selectedGrades | Number(gradeFilter.key);
                }
            });

            if (selectedGrades === GradeLevels.GradeLevels.NA) {
                selectedGrades = GradeLevels.GradeLevels.All;
            }

            itemCards = itemCards.filter(g => GradeLevels.contains(selectedGrades, g.grade));
        }

        if (subjects && subjects.filterOptions) {
            let subjectCode:string|undefined;

            subjects.filterOptions.forEach(s => {
                if(s.isSelected){
                    subjectCode = s.key;
                }
            });
            if(subjectCode !== undefined){
                itemCards = itemCards.filter(i => subjectCode === i.subjectCode);
            }
        }
        //TODO: What is CAT technology? Filter? Ignore?
        if (techTypes && techTypes.filterOptions) {
            if (techTypes.filterOptions.find(t => t.key === "PT" && t.isSelected)) {
                itemCards = itemCards.filter(i => i.isPerformanceItem);
            } else if (techTypes.filterOptions.find(t => t.key === "CAT" && t.isSelected)) {
                itemCards = itemCards.filter(i => !i.isPerformanceItem);
            }
        }

        return itemCards;
    }

    static updateUrl(filter: AdvancedFilterCategory[]) {
        let pairs: string[] = [];

        const grades = filter.find(afc => afc.label === "Grade");
        const subjects = filter.find(afc => afc.label === "Subjects");
        const techTypes = filter.find(afc => afc.label === "TechType");

        if (grades && !grades.disabled) {
            const gradeString: string[] = [];

            grades.filterOptions.forEach(g => {
                if (g.isSelected) {
                    gradeString.push(g.key.toString());
                }
            });

            if (gradeString.length !== 0) {
                gradeString.join(',');
                pairs.push("gradeLevels=" + gradeString);
            }
        }
        if (subjects && !subjects.disabled) {
            const subjString: string[] = [];

            subjects.filterOptions.forEach(s => {
                if (s.isSelected) {
                    subjString.push(s.key);
                }
            });

            if (subjString.length !== 0) {
                subjString.join(',');
                pairs.push("subjects=" + subjString);
            }
        }
        if (techTypes && !techTypes.disabled) {
            const techTypesString: string[] = [];

            techTypes.filterOptions.forEach(t => {
                if (t.isSelected) {
                    techTypesString.push(t.key);
                }
            });
            if (techTypesString.length !== 0) {
                techTypesString.join(',');
                pairs.push("techTypes=" + techTypesString);
            }
        }

        let query: string;
        if (pairs.length === 0) {
            query = "/";
        } else {
            query = "?" + pairs.join("&");
        }
        history.replaceState(null, "", query);
    }

    static readUrl(filterOptions: AdvancedFilters) {
        const queryObject = parseQueryString(window.location.href);
        const subjects = queryObject["subjects"]
            ? queryObject["subjects"]!.map(
                subjCode => {
                    let subject = filterOptions.subjects.filterOptions.find(s => s.key === subjCode);
                    if (subject) {
                        return {
                            code: subject.key,
                            label: subject.label
                        }
                    }
                }) : [];

        const grades = queryObject["grades"]
            ? queryObject["grades"]!
                .map(gradeCode => Number(gradeCode) as GradeLevels.GradeLevels)
            : [];

        const techTypes = queryObject["techTypes"]
            ? queryObject["techTypes"]!.map(typeCode => {
                let techType = filterOptions.techTypes.filterOptions.find(t => t.key === typeCode);
                if (techType) {
                    return {
                        code: techType.key,
                        label: techType.label
                    }
                }
            }) : [];

        return {
            subjects: subjects,
            grades: grades,
            techTypes: techTypes
        } as ItemFilter;
    }
}