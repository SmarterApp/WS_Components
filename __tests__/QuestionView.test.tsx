import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { QuestionView } from '../QuestionView';
import * as Models from "../../Models";
import { configure, shallow } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

const itemView: Models.ItemView = {
    id: "12334",
    html: null,
    picturePath: null,
    captured: true,
    type: Models.ViewType.html
}

const question: Models.Question = {
    id: "something",
    view: itemView,
    data: null,
    questionNumber: null
}

const props  = {question}

describe("QuestionDataTable", () => {
    it("matches snapshot", () => {
        expect(shallow(<QuestionView {...props} />)).toMatchSnapshot();
    })
})