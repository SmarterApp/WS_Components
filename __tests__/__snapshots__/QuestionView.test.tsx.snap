// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionDataTable matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <QuestionView
    question={
      Object {
        "data": null,
        "id": "something",
        "questionNumber": null,
        "view": Object {
          "captured": true,
          "html": null,
          "id": "12334",
          "picturePath": null,
          "type": 1,
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Question #
        </h2>,
        null,
        null,
        <ItemViewComponent
          view={
            Object {
              "captured": true,
              "html": null,
              "id": "12334",
              "picturePath": null,
              "type": 1,
            }
          }
        />,
        null,
        undefined,
      ],
      "className": "item",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Question #",
            null,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Question #",
          null,
        ],
        "type": "h2",
      },
      null,
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "view": Object {
            "captured": true,
            "html": null,
            "id": "12334",
            "picturePath": null,
            "type": 1,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      null,
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Question #
          </h2>,
          null,
          null,
          <ItemViewComponent
            view={
              Object {
                "captured": true,
                "html": null,
                "id": "12334",
                "picturePath": null,
                "type": 1,
              }
            }
          />,
          null,
          undefined,
        ],
        "className": "item",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Question #",
              null,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Question #",
            null,
          ],
          "type": "h2",
        },
        null,
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "view": Object {
              "captured": true,
              "html": null,
              "id": "12334",
              "picturePath": null,
              "type": 1,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
